File: CSES/3. Dynamic Programming/CSES-1634_Minimizing Coins.py
"""
    LeetCode 322. Coin Change
    1. 枚舉硬幣 -> 不排序也可以，但排序後可以更快
    2. 枚舉金額 -> TLE
"""

n, x = map(int, input().split())
coins = list(map(int, input().split()))
coins.sort() # 不排序不會影響答案，但排序後可以更快

# dp[i] 表示湊成 i 元的最小硬幣數
dp = [float("inf")] * (x + 1)
dp[0] = 0
for c in coins:
    for j in range(c, x + 1): # dp[j] 表示湊成 j 元的最小硬幣數，可以從 j - c 轉換過來
        dp[j] = min(dp[j], dp[j - c] + 1)
print(dp[x] if dp[x] != float("inf") else -1)

File: CSES/3. Dynamic Programming/CSES-1634_Minimizing Coins.cpp
#include <bits/stdc++.h>
using namespace std;
const int INF = 0x3f3f3f3f;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, x;
    cin >> n >> x;
    vector<int> coins(n);
    for (int i = 0; i < n; i++) cin >> coins[i];
    sort(coins.begin(), coins.end()); // Not necessary, but can be faster
    vector<int> dp(x + 1, INF);
    dp[0] = 0;
    for (int i = 0; i < n; i++) {
        for (int j = coins[i]; j <= x; j++) {
            dp[j] = min(dp[j], dp[j - coins[i]] + 1);
        }
    }
    cout << (dp[x] != INF ? dp[x] : -1) << endl;
    return 0;
}